Raft follower replicates snapshot from leader

Restore follower partition service and group, and use a group reference to manage sub services for each follower partitions, and replicating the leader's snapshots.

Related to #876 and #877

Possible race condition in `SnapshotReplicationServiceTest#shouldRetryOnTopologyError` on line [137](https://github.com/zeebe-io/zeebe/blob/master/broker-core/src/test/java/io/zeebe/broker/clustering/base/snapshots/SnapshotReplicationServiceTest.java#L137).
You need to do this in the `SnapshotReplicationService` actor otherwise it can interfere with the `SnapshotReplicationService` execution actor thread and the main thread which runs the test method.

Mockito will have problems to change the mocks at sometime and then it will fail like this:

```java
08:41:48 [INFO] Running io.zeebe.broker.clustering.base.snapshots.SnapshotReplicationServiceTest
08:41:49 [ERROR] Tests run: 6, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.363 s <<< FAILURE! - in io.zeebe.broker.clustering.base.snapshots.SnapshotReplicationServiceTest
08:41:49 [ERROR] shouldRetryOnTopologyError(io.zeebe.broker.clustering.base.snapshots.SnapshotReplicationServiceTest)  Time elapsed: 0.014 s  <<< ERROR!
08:41:49 org.mockito.exceptions.misusing.UnfinishedStubbingException: 
08:41:49 
08:41:49 Unfinished stubbing detected here:
08:41:49 -> at io.zeebe.broker.clustering.base.snapshots.SnapshotReplicationServiceTest.shouldRetryOnTopologyError(SnapshotReplicationServiceTest.java:137)
08:41:49 
08:41:49 E.g. thenReturn() may be missing.
08:41:49 Examples of correct stubbing:
08:41:49     when(mock.isOk()).thenReturn(true);
08:41:49     when(mock.isOk()).thenThrow(exception);
08:41:49     doThrow(exception).when(mock).someVoidMethod();
08:41:49 Hints:
08:41:49  1. missing thenReturn()
08:41:49  2. you are trying to stub a final method, which is not supported
08:41:49  3: you are stubbing the behaviour of another mock inside before 'thenReturn' instruction if completed
08:41:49 
08:41:49 	at io.zeebe.broker.clustering.base.snapshots.SnapshotReplicationServiceTest.shouldRetryOnTopologyError(SnapshotReplicationServiceTest.java:137)
08:41:49 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
08:41:49 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
08:41:49 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
08:41:49 	at java.lang.reflect.Method.invoke(Method.java:498)
08:41:49 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
08:41:49 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
08:41:49 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
08:41:49 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
08:41:49 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
08:41:49 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
08:41:49 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
08:41:49 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
08:41:49 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
08:41:49 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
08:41:49 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
08:41:49 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
08:41:49 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
08:41:49 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
08:41:49 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
08:41:49 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
08:41:49 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
08:41:49 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
08:41:49 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
08:41:49 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
08:41:49 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
08:41:49 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
08:41:49 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
08:41:49 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
08:41:49 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)
08:41:49 
```

Corresponding build https://hq2.camunda.com/jenkins/zeebe/job/zeebe-DISTRO-maven-deploy/1301/consoleFull

Also there is a possible NullPointer on `#shouldAbortCurrentSnapshotIfOneChunkFailsAndMoveToNextOne` see failing test build
https://hq2.camunda.com/jenkins/zeebe/job/zeebe-DISTRO-maven-deploy/1303/consoleFull

Maybe you can have a look into this.

--- a/broker-core/src/main/java/io/zeebe/broker/clustering/api/SnapshotReplicationRequestHandler.java
+++ b/broker-core/src/main/java/io/zeebe/broker/clustering/api/SnapshotReplicationRequestHandler.java
@@ -76,7 +76,6 @@ public class SnapshotReplicationRequestHandler {
   Supplier<BufferWriter> handleListSnapshotsAsync(
       final DirectBuffer buffer, final int offset, final int length) {
     listSnapshotsRequest.wrap(buffer, offset, length);
-    listSnapshotsResponse.reset();
 
     final int partitionId = listSnapshotsRequest.getPartitionId();
     final Partition partition = trackedPartitions.get(partitionId);
@@ -109,6 +108,8 @@ public class SnapshotReplicationRequestHandler {
 
   private BufferWriter handleListSnapshots(final SnapshotStorage storage) {
     final List<SnapshotMetadata> snapshots = storage.listSnapshots();
+
+    listSnapshotsResponse.reset();
     for (final SnapshotMetadata snapshot : snapshots) {
       if (snapshot.isReplicable()) {
         listSnapshotsResponse.addSnapshot(

--- a/broker-core/src/main/java/io/zeebe/broker/clustering/base/snapshots/SnapshotReplicationService.java
+++ b/broker-core/src/main/java/io/zeebe/broker/clustering/base/snapshots/SnapshotReplicationService.java
@@ -190,7 +190,7 @@ public class SnapshotReplicationService extends Actor
   private void handleListSnapshotsResponse(final DirectBuffer buffer) {
     if (isErrorResponse(buffer)) {
       logErrorResponse("Error listing snapshots", buffer);
-      actor.runDelayed(ERROR_RETRY_INTERVAL, this::pollSnapshots);
+      actor.runDelayed(ERROR_RETRY_INTERVAL, this::pollLeaderForSnapshots);
       return;
     }
 

--- a/broker-core/src/test/java/io/zeebe/broker/clustering/base/snapshots/SnapshotReplicationServiceTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/clustering/base/snapshots/SnapshotReplicationServiceTest.java
@@ -27,6 +27,7 @@ import io.zeebe.broker.clustering.base.partitions.Partition;
 import io.zeebe.broker.clustering.base.topology.NodeInfo;
 import io.zeebe.broker.clustering.base.topology.PartitionInfo;
 import io.zeebe.broker.util.BufferingClientOutput;
+import io.zeebe.broker.util.BufferingClientOutput.Request;
 import io.zeebe.broker.util.ControlledTopologyManager;
 import io.zeebe.clustering.management.ErrorResponseCode;
 import io.zeebe.clustering.management.FetchSnapshotChunkRequestEncoder;
@@ -186,6 +187,34 @@ public class SnapshotReplicationServiceTest {
     assertThat(output.getSentRequests().size()).isEqualTo(2);
     assertThat(output.getLastRequest().getTemplateId())
         .isEqualTo(ListSnapshotsRequestEncoder.TEMPLATE_ID);
+  }
+
+  @Test
+  public void shouldRefreshLeaderAddressOnListSnapshotsError() throws Exception {
+    // given
+    final ErrorResponse response =
+        new ErrorResponse().setCode(ErrorResponseCode.PARTITION_NOT_FOUND).setData("fail");
+
+    // when
+    installService();
+
+    // then
+    assertThat(output.getSentRequests().size()).isEqualTo(1);
+    assertThat(output.getLastRequest().getTemplateId())
+        .isEqualTo(ListSnapshotsRequestEncoder.TEMPLATE_ID);
+
+    // when
+    final NodeInfo newLeader = createNodeInfo("0.0.0.0", 12345);
+    topologyManager.setPartitionLeader(partition, newLeader);
+    output.getLastRequest().respondWith(response);
+    actorSchedulerRule.workUntilDone();
+    actorSchedulerRule.waitForTimer(SnapshotReplicationService.ERROR_RETRY_INTERVAL);
+
+    // then
+    assertThat(output.getSentRequests().size()).isEqualTo(2);
+    final Request lastRequest = output.getLastRequest();
+    assertThat(lastRequest.getTemplateId()).isEqualTo(ListSnapshotsRequestEncoder.TEMPLATE_ID);
+    verify(transport).registerRemoteAddress(newLeader.getManagementApiAddress());
   }
 
   @Test

--- a/util/src/main/java/io/zeebe/util/FileUtil.java
+++ b/util/src/main/java/io/zeebe/util/FileUtil.java
@@ -131,7 +131,7 @@ public class FileUtil {
   public static void replace(Path src, Path dest) throws IOException {
     try {
       Files.move(src, dest, ATOMIC_MOVE);
-    } catch (final FileAlreadyExistsException e) {
+    } catch (final Exception e) {
       // failed with atomic move, lets try again with normal replace move
       Files.move(src, dest, REPLACE_EXISTING);
     }

